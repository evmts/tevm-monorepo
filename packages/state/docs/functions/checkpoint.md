[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / checkpoint

# Function: checkpoint()

> **checkpoint**(`baseState`): () => `Promise`\<`void`\>

Checkpoints the current change-set to the instance since the
last call to checkpoint.

## Parameters

• **baseState**: [`BaseState`](../type-aliases/BaseState.md)

## Returns

`Function`

### Returns

`Promise`\<`void`\>

## Source

[packages/state/src/actions/checkpoint.js:6](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/actions/checkpoint.js#L6)
