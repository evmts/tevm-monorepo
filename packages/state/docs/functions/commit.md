[**@tevm/state**](../README.md) • **Docs**

***

[@tevm/state](../globals.md) / commit

# Function: commit()

> **commit**(`baseState`): (`createNewStateRoot`?) => `Promise`\<`void`\>

Commits the current change-set to the instance since the
last call to checkpoint.

## Parameters

• **baseState**: [`BaseState`](../type-aliases/BaseState.md)

## Returns

`Function`

Commits the current state.

### Parameters

• **createNewStateRoot?**: `boolean`

Whether to create a new state root
Defaults to true.
This api is not stable

### Returns

`Promise`\<`void`\>

## Source

[packages/state/src/actions/commit.js:11](https://github.com/evmts/tevm-monorepo/blob/main/packages/state/src/actions/commit.js#L11)
