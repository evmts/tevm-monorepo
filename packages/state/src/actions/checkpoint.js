/**
 * Checkpoints the current change-set to the instance since the
 * last call to checkpoint.
 * @type {import("../state-types/index.js").StateAction<'checkpoint'>}
 */
export const checkpoint = (baseState) => () => {
	baseState.caches.accounts.checkpoint()
	baseState.caches.storage.checkpoint()
	baseState.caches.contracts.checkpoint()
	baseState.logger.debug('checkpointed state!')
	return Promise.resolve()
}
