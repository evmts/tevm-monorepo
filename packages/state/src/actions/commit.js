import { keccak256, numberToHex, toHex } from 'viem'
import { dumpCanonicalGenesis } from './dumpCannonicalGenesis.js'

// TODO we might want to sometimes prune state roots

/**
 * Commits the current change-set to the instance since the
 * last call to checkpoint.
 * @type {import("../state-types/index.js").StateAction<'commit'>}
 */
export const commit =
	(baseState) =>
	async (createNewStateRoot = false) => {
		const state = await dumpCanonicalGenesis(baseState)()
		// This is kinda hacky we are just ranodmly generating new state roots
		// since we don't use a trie we are kinda hacking this
		// This might be needlessy slow as state gets big but for now it feels the most rubust wrt correctness

		const newStateRoot = (() => {
			if (!createNewStateRoot) {
				baseState.logger.debug('Comitting to existing state root...')
				return baseState.getCurrentStateRoot()
			}
			/**
			 * @type {import('../state-types/ParameterizedTevmState.js').ParameterizedTevmState}
			 */
			const jsonSerializableState = {}

			for (const [k, v] of Object.entries(state)) {
				const { nonce, balance, storageRoot, codeHash, deployedBytecode } = v
				jsonSerializableState[k] = {
					...v,
					nonce: numberToHex(nonce),
					balance: numberToHex(balance),
					storageRoot: storageRoot,
					codeHash: codeHash,
					...(deployedBytecode !== undefined ? { deployedBytecode } : {}),
				}
			}
			const root = keccak256(toHex(JSON.stringify(jsonSerializableState)))
			baseState.logger.debug({ root }, 'Committing to new state root...')
			return root
		})()

		baseState.stateRoots.set(newStateRoot, state)
		baseState.logger.debug({ stateRoot: newStateRoot }, 'Saving state roots...')
		baseState.setCurrentStateRoot(newStateRoot)

		baseState.caches.accounts.commit()
		baseState.caches.contracts.commit()
		baseState.caches.storage.commit()

		baseState.logger.debug('State roots are comitted successfully')

		baseState.options.onCommit?.(baseState)

		return
	}
