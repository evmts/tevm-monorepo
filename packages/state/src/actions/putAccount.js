/**
 * Saves an account into state under the provided `address`.
 * @type {import("../state-types/index.js").StateAction<'putAccount'>}
 */
export const putAccount = (baseState) => async (address, account) => {
	if (account !== undefined) {
		baseState.caches.accounts?.put(address, account)
	} else {
		baseState.caches.accounts?.del(address)
	}
}
