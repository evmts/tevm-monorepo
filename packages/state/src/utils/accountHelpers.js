import { Rlp } from '@tevm/rlp'
import { createAccount, createAddressFromString as createAddress } from '@tevm/utils'

/**
 * Helper to create an address from a string
 * @param {string} addressString - The address string
 * @returns {import('@tevm/utils').EthjsAddress} The address
 */
export function createAddressFromString(addressString) {
	return createAddress(addressString)
}

/**
 * Creates an account from RLP serialized data
 * @param {Uint8Array} serialized - The RLP serialized account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromRlpSerializedAccount(serialized) {
	const decoded = Rlp.decode(serialized)
	if (!Array.isArray(decoded) || decoded.length !== 4) {
		throw new Error('Invalid RLP serialized account')
	}

	const [nonce, balance, storageRoot, codeHash] = decoded

	/** @type {{nonce: any, balance: any, storageRoot: any, codeHash: any}} */
	const accountData = {
		nonce,
		balance,
		storageRoot,
		codeHash,
	}

	return createAccount(accountData)
}

/**
 * Creates an account from account data
 * @param {{nonce?: bigint, balance?: bigint, storageRoot?: Uint8Array, codeHash?: Uint8Array}} accountData - The account data
 * @returns {import('@tevm/utils').EthjsAccount} The account
 */
export function fromAccountData(accountData) {
	return createAccount(accountData)
}
