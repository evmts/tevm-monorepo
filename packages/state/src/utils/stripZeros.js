import { InternalError } from '@tevm/errors'

/**
 * @param {Uint8Array} bytes
 * @returns {Uint8Array}
 */
export const stripZeros = (bytes) => {
	if (!(bytes instanceof Uint8Array)) {
		throw new InternalError('Unexpected type')
	}
	return bytes.slice(bytes.findIndex(/** @param {number} entry*/ (entry) => entry !== 0))
}
