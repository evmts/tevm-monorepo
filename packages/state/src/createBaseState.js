import { AccountCache, CacheType, StorageCache } from '@ethereumjs/statemanager'
import { createLogger } from '@tevm/logger'
import { ContractCache } from './ContractCache.js'
import { checkpoint } from './actions/checkpoint.js'
import { commit } from './actions/commit.js'
import { generateCanonicalGenesis } from './actions/generateCannonicalGenesis.js'

/**
 * @type {import('viem').Hex}
 */
const INITIAL_STATE_ROOT = '0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421'

/**
 * @internal
 * Creates the core data structure for ethereum state
 * @param {import('./state-types/index.js').StateOptions} options
 * @returns {import('./BaseState.js').BaseState}
 * Creates the core data structure the state manager operates on
 */
export const createBaseState = (options) => {
	const logger = createLogger({
		level: options.loggingLevel ?? 'warn',
		name: '@tevm/state-manager',
	})
	/**
	 * @type {import('./state-types/StateRoots.js').StateRoots}
	 */
	const stateRoots = options.stateRoots ?? new Map()
	if (options.genesisState && options.currentStateRoot) {
		stateRoots.set(options.currentStateRoot, options.genesisState)
	} else {
		stateRoots.set(INITIAL_STATE_ROOT, options.genesisState ?? {})
	}
	let currentStateRoot = options.currentStateRoot ?? INITIAL_STATE_ROOT
	/**
	 * @type {import('./BaseState.js').BaseState}
	 */
	const state = {
		logger,
		getCurrentStateRoot: () => currentStateRoot,
		setCurrentStateRoot: (root) => {
			if (!state.stateRoots.has(root)) {
				throw new Error('Cannot set state root to non existing state root')
			}
			currentStateRoot = root
		},
		stateRoots,
		options,
		caches: {
			contracts: new ContractCache(),
			accounts: new AccountCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
			storage: new StorageCache({
				size: 100_000,
				type: CacheType.ORDERED_MAP,
			}),
		},
		ready: () => genesisPromise.then(() => true),
	}
	const genesisPromise = (
		options.genesisState !== undefined && options.currentStateRoot === undefined
			? generateCanonicalGenesis(state)(options.genesisState)
			: Promise.resolve().then(() => {
					if (options.currentStateRoot) {
						state.setCurrentStateRoot(options.currentStateRoot)
						return generateCanonicalGenesis(state)(options.genesisState ?? stateRoots.get(options.currentStateRoot))
					}
					return Promise.resolve()
				})
	).then(async () => {
		await checkpoint(state)()
		await commit(state)()
		logger.debug('StateManager is ready')
	})

	return state
}
