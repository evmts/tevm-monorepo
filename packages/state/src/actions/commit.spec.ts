import { beforeEach, describe, expect, it, jest } from 'bun:test'
import { createAddress } from '@tevm/address'
import { EthjsAccount } from '@tevm/utils'
import { createBaseState } from '../createBaseState.js'
import { checkpoint } from './checkpoint.js'
import { commit } from './commit.js'
import { putAccount } from './putAccount.js'

describe(commit.name, () => {
	it('should clear all storage entries for the account corresponding to `address`', async () => {
		// No mocks
		const baseState = createBaseState({
			loggingLevel: 'warn',
		})
		await putAccount(baseState)(createAddress(69696969), EthjsAccount.fromAccountData({ balance: 20n }))
		expect(baseState.caches.storage._checkpoints).toBe(0)
		expect(baseState.caches.accounts._checkpoints).toBe(0)
		expect(baseState.caches.contracts._checkpoints).toBe(0)
		await checkpoint(baseState)()
		expect(baseState.caches.storage._checkpoints).toBe(1)
		expect(baseState.caches.accounts._checkpoints).toBe(1)
		expect(baseState.caches.contracts._checkpoints).toBe(1)
		expect(baseState.caches.accounts._diffCache).toEqual([
			new Map(
				Object.entries({
					'0000000000000000000000000000000004277dc9': undefined,
				}),
			),
			new Map(),
		])
		expect(baseState.caches.storage._diffCache).toEqual([new Map(), new Map()])
		await commit(baseState)(true)
		expect(baseState.caches.storage._checkpoints).toBe(1)
		expect(baseState.caches.accounts._checkpoints).toBe(1)
		expect(baseState.caches.contracts._checkpoints).toBe(1)
		expect(baseState.caches.accounts._diffCache).toEqual([
			new Map(
				Object.entries({
					'0000000000000000000000000000000004277dc9': undefined,
				}),
			),
			new Map(),
		])
		expect(baseState.caches.storage._diffCache).toEqual([new Map(), new Map()])
		expect(baseState.getCurrentStateRoot()).toEqual(
			'0x886f43e0144bf4f5748e999d0178ed7e4edea8ad708e0bf26a61341e8ae91d1e',
		)
	})

	let baseState: ReturnType<typeof createBaseState>

	beforeEach(() => {
		baseState = createBaseState({
			loggingLevel: 'warn',
		})
		baseState.getCurrentStateRoot = jest.fn(() => 'existingStateRoot') as any
		baseState.setCurrentStateRoot = jest.fn()
		baseState.logger.debug = jest.fn()
		baseState.options.onCommit = jest.fn()
		baseState.stateRoots.set = jest.fn()
		baseState.caches.accounts.commit = jest.fn()
		baseState.caches.contracts.commit = jest.fn()
		baseState.caches.storage.commit = jest.fn()
	})

	it('should commit to existing state root', async () => {
		await checkpoint(baseState)()
		await commit(baseState)()

		expect(baseState.getCurrentStateRoot).toHaveBeenCalled()
		expect(baseState.setCurrentStateRoot).toHaveBeenCalledWith('existingStateRoot')
		expect(baseState.logger.debug).toHaveBeenCalledWith('Comitting to existing state root...')
		expect(baseState.stateRoots.set).toHaveBeenCalledWith('existingStateRoot', expect.any(Object))
		expect(baseState.caches.accounts.commit).toHaveBeenCalled()
		expect(baseState.caches.contracts.commit).toHaveBeenCalled()
		expect(baseState.caches.storage.commit).toHaveBeenCalled()
		expect(baseState.options.onCommit).toHaveBeenCalledWith(baseState)
	})

	it('should commit to a new state root', async () => {
		await checkpoint(baseState)()
		await commit(baseState)(true)

		expect(baseState.setCurrentStateRoot).toHaveBeenCalledWith(expect.any(String))
		expect(baseState.logger.debug).toHaveBeenCalledWith(
			expect.objectContaining({ root: expect.any(String) }),
			'Committing to new state root...',
		)
		expect(baseState.stateRoots.set).toHaveBeenCalledWith(expect.any(String), expect.any(Object))
		expect(baseState.caches.accounts.commit).toHaveBeenCalled()
		expect(baseState.caches.contracts.commit).toHaveBeenCalled()
		expect(baseState.caches.storage.commit).toHaveBeenCalled()
		expect(baseState.options.onCommit).toHaveBeenCalledWith(baseState)
	})

	it('should handle onCommit callback correctly', async () => {
		const onCommitMock = jest.fn()
		baseState.options.onCommit = onCommitMock

		await checkpoint(baseState)()
		await commit(baseState)()

		expect(onCommitMock).toHaveBeenCalledWith(baseState)
	})
})
