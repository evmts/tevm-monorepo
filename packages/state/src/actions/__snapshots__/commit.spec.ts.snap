// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`commit should clear all storage entries for the account corresponding to `address` 1`] = `
[
  
Map {
    "0000000000000000000000000000000004277dc9" => undefined,
  }
,
  Map {},
]
`;

exports[`commit should clear all storage entries for the account corresponding to `address` 2`] = `
[
  
Map {
    "0000000000000000000000000000000004277dc9" => undefined,
  }
,
  Map {},
]
`;

exports[`commit should clear all storage entries for the account corresponding to `address` 3`] = `
[
  Map {},
  Map {},
]
`;

exports[`commit should clear all storage entries for the account corresponding to `address` 4`] = `"0x886f43e0144bf4f5748e999d0178ed7e4edea8ad708e0bf26a61341e8ae91d1e"`;
